catroid/src/main/java/org/catrobat/catroid/embroidery/DSTFileGenerator.java

```diff
@@ -0,0 +1,317 @@
+		header += formatString("+X:%-4d\n" + substituteChar, (int) maxX);
+		header += formatString("-X:%-4d\n" + substituteChar, (int) minX);
+		header += formatString("+Y:%-4d\n" + substituteChar, (int) maxY);
+		header += formatString("-Y:%-4d\n" + substituteChar, (int) minY);
+		header += formatString("AX:%-5d\n" + substituteChar, (int) (lastStitch.x - firstStitch.x));
+		header += formatString("AY:%-5d\n" + substituteChar, (int) (lastStitch.y - firstStitch.y));
+		header += formatString("MX:%-5d\n" + substituteChar, mx);
+		header += formatString("MY:%-5d\n" + substituteChar, my);
+		header += String.format("PD:%-5s\n" + substituteChar, pd).replace(' ', '\0');
+
+		return header;
+	}
+
+	private String formatString(String format, int value) {
+		return String.format(format, value).replace(' ', '\0');
+	}
+
+	private void writeStitches() throws IllegalArgumentException, IOException {
+		int xChange;
+		int yChange;
+
+		ListIterator<PointF> iterator = stitchPoints.listIterator();
+		PointF currentPoint = stitchPoints.get(0);
+		while (iterator.hasNext()) {
+			byte0 = 0;
+			byte1 = 0;
+			byte2 = 0x03;
+
+			PointF nextPoint = iterator.next();
+			xChange = (int) (nextPoint.x - currentPoint.x);
+			yChange = (int) (nextPoint.y - currentPoint.y);
+
+			setBitsForStitch(xChange, true);
+			setBitsForStitch(yChange, false);
+
+			writeBytes();
+			currentPoint = nextPoint;
+		}
+	}
+
+	private void setBitsForStitch(int absoluteChange, boolean isXCoord) throws IllegalArgumentException {
+		final int[] bitsXCoord = {2, 3, 0, 1};
+		final int[] bitsYCoord = {5, 4, 7, 6};
+
+		int[] bitsToSet = bitsYCoord;
+
+		if (isXCoord) {
+			bitsToSet = bitsXCoord;
+		}
+
+		// CHECKSTYLE DISABLE OneStatementPerLineCheck FOR 11 LINES
+		// CHECKSTYLE DISABLE LeftCurlyCheck FOR 12 LINES
+		if (absoluteChange >= +41) { byte2 += setBit(bitsToSet[0]); absoluteChange -= 81; }
+		if (absoluteChange <= -41) { byte2 += setBit(bitsToSet[1]); absoluteChange += 81; }
+		if (absoluteChange >= +14) { byte1 += setBit(bitsToSet[0]); absoluteChange -= 27; }
+		if (absoluteChange <= -14) { byte1 += setBit(bitsToSet[1]); absoluteChange += 27; }
+		if (absoluteChange >= +5) { byte0 += setBit(bitsToSet[0]); absoluteChange -= 9; }
+		if (absoluteChange <= -5) { byte0 += setBit(bitsToSet[1]); absoluteChange += 9; }
+		if (absoluteChange >= +2) { byte1 += setBit(bitsToSet[2]); absoluteChange -= 3; }
+		if (absoluteChange <= -2) { byte1 += setBit(bitsToSet[3]); absoluteChange += 3; }
+		if (absoluteChange >= +1) { byte0 += setBit(bitsToSet[2]); absoluteChange -= 1; }
+		if (absoluteChange <= -1) { byte0 += setBit(bitsToSet[3]); absoluteChange += 1; }
+		if (absoluteChange != 0) {
+			throw new IllegalArgumentException("Remaining stitch distance should be zero");
+		}
+	}
+
+	private void writeEndFile() throws IOException {
+		byte0 = 0;
+		byte1 = 0;
+		byte2 = 0xF3;
+
+		writeBytes();
+		fileStream.flush();
+		fileStream.close();
+	}
+
+	private void writeBytes() throws IOException {
```

Comment:
I would prefer to call fileStream.write directly with the bytes you want to write and omit the member variables